e2e/nomostest/nt.go

```diff
@@ -749,30 +757,63 @@ func (nt *NT) portForwardGitServer() {
 
 // portForwardRegistryServer forwards the registry-server deployment to a port.
 func (nt *NT) portForwardRegistryServer() {
-	provider := nt.OCIProvider.(*registryproviders.LocalOCIProvider)
-	resetImages := func(newPort int, podName string) {
-		for _, image := range nt.ociImages {
-			// construct push URL with provided port. Calling LocalPort would lead to deadlock
-			if err := image.Push(provider.PushURLWithPort(newPort, image.Name)); err != nil {
-				nt.T.Fatalf("pushing image: %w", err)
+	var onReadyCallbacks []func(int, string)
+	if *e2e.OCIProvider == e2e.Local {
+		provider := nt.OCIProvider.(*registryproviders.LocalOCIProvider)
+		resetImages := func(newPort int, podName string) {
+			for _, image := range nt.ociImages {
+				// construct push URL with provided port. Calling LocalPort would lead to deadlock
+				if err := image.Push(provider.PushURLWithPort(newPort, image.Name)); err != nil {
+					nt.T.Fatalf("pushing image: %w", err)
+				}
 			}
 		}
+		onReadyCallbacks = append(onReadyCallbacks, resetImages)
+	}
+	if *e2e.HelmProvider == e2e.Local {
+		provider := nt.HelmProvider.(*registryproviders.LocalHelmProvider)
+		resetPackages := func(newPort int, podName string) {
+			for _, pkg := range nt.helmPackages {
+				// construct push URL with provided port. Calling LocalPort would lead to deadlock
+				if err := pkg.Push(provider.PushURLWithPort(newPort, pkg.Name)); err != nil {
+					nt.T.Fatalf("pushing image: %w", err)
+				}
+			}
+		}
+		onReadyCallbacks = append(onReadyCallbacks, resetPackages)
 	}
-	nt.T.Cleanup(func() {
-		// clear PortForwarder after each test. at this point it has been stopped
-		// a new PortForwarder is created for each test.
-		provider.PortForwarder = nil
-	})
-	provider.PortForwarder = nt.newPortForwarder(
+	portForwarder := nt.newPortForwarder(
 		TestRegistryNamespace,
 		TestRegistryServer,
 		fmt.Sprintf(":%d", RegistryHTTPPort),
-		portforwarder.WithOnReadyCallback(resetImages),
+		portforwarder.WithOnReadyCallback(func(newPort int, podName string) {
```

Comment:
Is there a possible race condition with the tests trying to use these images & charts before they've all been pushed?

Do we need to block waiting for them all to be pushed before continuing? If so, should we just do it inline here without the callbacks?

Reply:
This callback is called from a different thread by the portforwarder. The point of this is that the registry could crash at any time during the test, and this callback handles the auto-recovery. I don't think there's much to do about synchronizing this in the general case.

Reply:
Does the proxy block/delay requests if the backend is down?

If so, does that blocking also include the time spent re-pushing the images that are expected to exist?

Reply:
In a way. The [portforwarder blocks calls to LocalPort while the backend is down](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/portforwarder/port_forwarder.go#L142), which is how the test scaffolding gets the proxy port. Any test logic that uses `LocalPort` may need some retry logic to recover if the backend crashes while trying to get the LocalPort.

Before unlocking LocalPort, the portforwarder [invokes each of the registered callbacks](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/portforwarder/port_forwarder.go#L151), which in this case re-pushes the images which are expected to exist.

Reply:
Ah. I see `update` sets `isReady` first, but then calls `onReadyCallback` inside the same lock.

That should prevent the race condition I was concerned about.

It's messy to have the callback impl calling a PortForwarder method inside the callback tho. I wonder if we could separate the building of the URL path from the building of the address. Building the address needs to be locked but building the path doesn't. Then we could just pass the address into the callback, instead of calling a method to do it. And adding the path could be a package function, instead of a struct method.

Reply:
You or I could do another PR on this after. Doesn't need to block this one.